platform/security/idp/security-idp-plugin/security-idp-plugin-nameidpolicy/src/main/java/org/codice/ddf/security/idp/plugin/nameidpolicy/NameIdPolicyPresignPlugin.java

```diff
@@ -0,0 +1,135 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(NameIdPolicyPresignPlugin.class);
+
+  private static final String EMAIL_ATTRIBUTE_NAME =
+      "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress";
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+    NameIDPolicy nameIdPolicy = authnRequest.getNameIDPolicy();
+
+    if (nameIdPolicy == null
+        || (StringUtils.isEmpty(nameIdPolicy.getFormat())
+            && StringUtils.isEmpty(nameIdPolicy.getSPNameQualifier()))) {
+      return;
+    }
+
+    response
+        .getAssertions()
+        .stream()
+        .map(assertion -> assertion.getSubject().getNameID())
+        .forEach(
+            assertionNameId ->
+                setAssertionNameIdQualifierAndFormat(
+                    assertionNameId,
+                    nameIdPolicy.getFormat(),
+                    nameIdPolicy.getSPNameQualifier(),
+                    response));
+  }
+
+  private void setAssertionNameIdQualifierAndFormat(
+      NameID assertionNameId,
+      @Nullable String nameIdFormatPolicy,
+      @Nullable String spNameQualifierPolicy,
+      Response response) {
+
+    if (StringUtils.isNotEmpty(spNameQualifierPolicy)) {
+      assertionNameId.setSPNameQualifier(spNameQualifierPolicy);
+    }
+
+    if (StringUtils.isNotEmpty(nameIdFormatPolicy)) {
+      switch (nameIdFormatPolicy) {
+          // supported NameIDFormats
+        case NameID.UNSPECIFIED:
+          return; // avoid changing the Format later
+        case NameID.PERSISTENT:
+          // TODO DDF-3965
+          break;
+
+          // partially supported NameIDFormats
+        case NameID.X509_SUBJECT:
+          LOGGER.warn("Specifying the \"X509Subject\" NameIDPolicy Format is not fully supported.");
+          if (!NameID.X509_SUBJECT.equals(assertionNameId.getFormat())) {
+            throw new UnsupportedOperationException(
+                "The \"X509Subject\" NameID could not be retrieved.");
+          }
+          break;
+        case NameID.EMAIL:
+          LOGGER.warn("Specifying the \"Email\" NameIDPolicy Format is not fully supported.");
+          assertionNameId.setValue(resolveEmail(response));
+          break;
+
+          // not supported NameIDFormats
+        case NameID.TRANSIENT:
+        case NameID.WIN_DOMAIN_QUALIFIED:
+        case NameID.KERBEROS:
+        case NameID.ENTITY:
+        default:
+          throw new UnsupportedOperationException(
+              String.format(
+                  "The NameIDPolicy Format of value %s is not supported.", nameIdFormatPolicy));
+      }
+      assertionNameId.setFormat(nameIdFormatPolicy);
+    }
+  }
+
+  private String resolveEmail(Response response) {
+    return response
+        .getAssertions()
+        .stream()
+        .flatMap(assertion -> assertion.getAttributeStatements().stream())
```

Comment:
✏️ Extracting these two `flatMap`s to a method would also help clarify the behavior here. Combined with the Predicate suggestion, you could end up with something like the following:
```java
  private String resolveEmail(Response response) {
    Predicate<Attribute> attributePredicate = attribute ->
        EMAIL_ATTRIBUTE_NAME.equals(attribute.getName())
            || NameID.EMAIL.equals(attribute.getNameFormat());
    
    return extractAssertionAttributes(response.getAssertions().stream())
        .filter(attributePredicate)
        .flatMap(attribute -> attribute.getAttributeValues().stream())
        .filter(attributeValue -> attributeValue instanceof XSString)
        .map(attributeValue -> (XSString) attributeValue)
        .map(XSString::getValue)
        .filter(StringUtils::isNotBlank)
        .findFirst()
        .orElseThrow(
            () ->
                new UnsupportedOperationException(
                    "The \"Email\" NameID could not be retrieved for this principal."));
  }

  private Stream<Attribute> extractAssertionAttributes(Stream<Assertion> stream) {
    return stream
        .flatMap(assertion -> assertion.getAttributeStatements().stream())
        .flatMap(statement -> statement.getAttributes().stream());
  }
```

Reply:
I'd still like to see this changed to be more compact and readable without level-switching while scanning the code.

Reply:
I'll see what I can do!

Reply:
Slightly better suggestion...change that `extractAssertionAttributes` method to take the `Response` instead.